### PR TITLE
[SPARK-29252][BUILD] Upgrade zookeeper to 3.4.14 and fix vulnerabilities.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
The current code uses org.apache.zookeeper:zookeeper:jar:3.4.6 and it will cause a security vulnerabilities. We could get some security info from https://www.tenable.com/cve/CVE-2019-0201

This reference remind to upgrate the version of `zookeeper` to 3.4.14 or later.

### Why are the changes needed?
This PR fix the security vulnerabilities.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Exists UT.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -2979,7 +2979,7 @@
       <properties>
         <hadoop.version>3.2.0</hadoop.version>
         <curator.version>2.13.0</curator.version>
-        <zookeeper.version>3.4.13</zookeeper.version>
```
Comment (dongjoon-hyun) [inline pom.xml]: Thank you, @beliefer . For this line, I'll include my Hadoop 3.2.1 PR, too.

File: pom.xml [hunk 2]
```diff
@@ -125,7 +125,7 @@
     <hadoop.version>2.7.4</hadoop.version>
     <protobuf.version>2.5.0</protobuf.version>
     <yarn.version>${hadoop.version}</yarn.version>
-    <zookeeper.version>3.4.6</zookeeper.version>
+    <zookeeper.version>3.4.14</zookeeper.version>
```
Comment (wangyum) [inline pom.xml]: Is it safe to upgrade `zookeeper` to `3.4.14` for Hadoop 2.7?

**Comment by beliefer** [comment]
Retest this please.

**Comment by beliefer** [comment]
Retest this please.

File: pom.xml (hunk 2 above)
Comment (srowen) [reply pom.xml]: Yes that's a good question @beliefer - what does 2.7 use?
I suspect we can update it. In that case we don't need to special-case the ZK version below anymore.

File: dev/deps/spark-deps-hadoop-2.7 [hunk 3]
```diff
@@ -20,6 +20,7 @@ arpack_combined_all-0.1.jar
 arrow-format-0.12.0.jar
 arrow-memory-0.12.0.jar
 arrow-vector-0.12.0.jar
+audience-annotations-0.5.0.jar
```
Comment (srowen) [inline dev/deps/spark-deps-hadoop-2.7]: This is an ALv2 dependency from Yetus, it seems. It needs a line in LICENSE-binary with other ASF dependencies. We'd also have to check if it has a NOTICE file bundled to append to NOTICE-binary

File: dev/deps/spark-deps-hadoop-3.2 [hunk 4]
```diff
@@ -211,6 +211,7 @@ spire-macros_2.12-0.17.0-M1.jar
 spire-platform_2.12-0.17.0-M1.jar
 spire-util_2.12-0.17.0-M1.jar
 spire_2.12-0.17.0-M1.jar
+spotbugs-annotations-3.1.9.jar
```
Comment (srowen) [inline dev/deps/spark-deps-hadoop-3.2]: Uh oh, this is LGPL. We can't include this. https://spotbugs.github.io/
It looks like ZK specifically excludes it from their binary release: https://github.com/apache/zookeeper/commit/372e713a9d2d9264417313e5d68e9437ffddd0f5
Therefore, we probably just need to write an exclusion rule to not pull in this dependency, as it sounds like it's not strictly needed at runtime (just annotations)

**Review by srowen — CHANGES_REQUESTED** [review]

File: dev/deps/spark-deps-hadoop-3.2 (hunk 4 above)
Comment (beliefer) [reply dev/deps/spark-deps-hadoop-3.2]: @srowen Thank you for your very detailed investigation. I learned more. I will try to exclude it.

File: dev/deps/spark-deps-hadoop-2.7 (hunk 3 above)
Comment (beliefer) [reply dev/deps/spark-deps-hadoop-2.7]: @srowen Thanks very much! However, I have not fully understood what you mean. Do we need NOTICE file?

File: pom.xml (hunk 2 above)
Comment (beliefer) [reply pom.xml]: Hadoop 2.7 uses `zookeeper:3.4.6` by default.
It's OK. I reference https://www.cnblogs.com/ywjfx/p/11344345.html.

Comment (wangyum) [reply pom.xml]: Could you try to upgrade zookeeper to 3.4.14 for [Hadoop 2.7.4](https://github.com/apache/hadoop/tree/release-2.7.4-RC0), and check if Hadoop test can pass?

File: dev/deps/spark-deps-hadoop-2.7 (hunk 3 above)
Comment (beliefer) [reply dev/deps/spark-deps-hadoop-2.7]: I checked the `audience-annotations-0.5.0.jar` and find the NOTICE file.
The content of the NOTICE file show below:

```
Apache Yetus - Audience Annotations
Copyright 2015-2017 The Apache Software Foundation

This product includes software developed at
The Apache Software Foundation (http://www.apache.org/).
```

Comment (srowen) [reply dev/deps/spark-deps-hadoop-2.7]: That's fine, just append this to the bottom of `NOTICE-binary`

Comment (beliefer) [reply dev/deps/spark-deps-hadoop-2.7]: OK. I will append the `NOTICE-binary`.

File: pom.xml [hunk 5]
```diff
@@ -2979,7 +2983,7 @@
       <properties>
         <hadoop.version>3.2.0</hadoop.version>
         <curator.version>2.13.0</curator.version>
-        <zookeeper.version>3.4.13</zookeeper.version>
+        <zookeeper.version>3.4.14</zookeeper.version>
```
Comment (srowen) [inline pom.xml]: I think you can just remove this if the profile does not change the value.

**Review by srowen — COMMENTED** [review]
Looks OK to me. We could probably exclude the other annotations too without issue, but it's OK to leave them as is.

File: pom.xml (hunk 5 above)
Comment (beliefer) [reply pom.xml]: OK. I will remove this line.

**Review by srowen — APPROVED** [review]

**Comment by beliefer** [comment]
@srowen Thanks for your help! @dongjoon-hyun @wangyum Thanks for all your review！

**Pull request closed without merging**